examples/advanced_operations/add_dynamic_page_feed.py

```diff
@@ -0,0 +1,295 @@
+import sys
+import uuid
+from datetime import datetime, timedelta
+from google.api_core import protobuf_helpers
+
+from google.ads.google_ads.client import GoogleAdsClient
+from google.ads.google_ads.errors import GoogleAdsException
+
+
+class FeedDetails(object):
+    def __init__(self, resource_name, url_attribute_id, label_attribute_id):
+        self.resource_name = resource_name
+        self.url_attribute_id = url_attribute_id
+        self.label_attribute_id = label_attribute_id
+
+def main(client, customer_id, campaign_id, ad_group_id):
+    dsa_page_url_label = 'discounts'
+
+    try:
+        feed_details = create_feed(client, customer_id)
+        create_feed_mapping(client, customer_id, feed_details)
+        create_feed_items(client, customer_id, feed_details, dsa_page_url_label)
+        update_campaign_dsa_setting(client, customer_id, campaign_id,
+                                    feed_details)
+        ad_group_resource_name = 'customers/{}/adGroups/{}'.format(
+            customer_id,ad_group_id)
+        add_dsa_targeting(client, customer_id, ad_group_resource_name,
+                          dsa_page_url_label)
+    except GoogleAdsException as ex:
+        print('Request with ID "{}" failed with status "{}" and includes the '
+              'following errors:'.format(ex.request_id, ex.error.code().name))
+        for error in ex.failure.errors:
+            print('\tError with message "{}".'.format(error.message))
+            if error.location:
+                for field_path_element in error.location.field_path_elements:
+                    print('\t\tOn field: {}'.format(
+                        field_path_element.field_name))
+        sys.exit(1)
+
+
+def create_feed(client, customer_id):
+    """Creates a page feed with URLs
+
+    Args:
+        client: an initialized GoogleAdsClient instance.
+        customer_id: a client customer ID str.
+
+    Returns:
+        A FeedDetails instance with information about the newly created feed.
+    """
+    feed_operation = client.get_type('FeedOperation', version='v1')
+    feed = feed_operation.create
+    feed.name.value = 'DSA Feed #{}'.format(uuid.uuid4())
+    feed.origin = client.get_type('FeedOriginEnum', version='v1').USER
+
+    feed_attribute_type_enum = client.get_type('FeedAttributeTypeEnum',
+                                               version='v1')
+
+    feed_attribute_url = feed.attributes.add()
+    feed_attribute_url.type = feed_attribute_type_enum.URL_LIST
+    feed_attribute_url.name.value = 'Page URL'
+
+    feed_attribute_label = feed.attributes.add()
+    feed_attribute_label.type = feed_attribute_type_enum.STRING_LIST
+    feed_attribute_label.name.value = 'Label'
+
+    feed_service = client.get_service('FeedService', version='v1')
+    response = feed_service.mutate_feeds(customer_id, [feed_operation])
+    feed_resource_name = response.results[0].resource_name
+
+    # We need to look up the attribute name and ID for the feed we just created
```

Comment:
Can this retrieval extracted into a function?

Reply:
Yep, done.